### PR TITLE
Add implementation of Exit command

### LINKED ISSUE #465: Add support for Exit command
#### Description:

Add implementation of Exit command added in https://github.com/BHoM/BHoM_Adapter/issues/302

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### NOTE: Depends on 

https://github.com/BHoM/BHoM_Adapter/pull/303

 ### Issues addressed by this PR

 Closes #465 

 

Adding support for the Exit command

 ### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/EjpTgU6jo_tJr30rbJs0a_gBnZnhI1rYcv3aQ-Yc3A6E2g?e=fenJ0J

 ### Changelog

 ### Additional comments

### CODE REVIEW

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
Minor check otherwise works well, even without a project loaded.

**Comment by peterjamesnugent** [comment]
Forgot to add - I had up to update the .NET Framework to 4.7.2 to get it to build.

File: Robot_Adapter/AdapterActions/Execute.cs [hunk 1]
```diff
@@ -101,6 +101,20 @@ public bool RunCommand(Analyse command)
 
         /***************************************************/
 
+        public bool RunCommand(Exit command)
+        {
+            if (command.SaveBeforeClose)
+                m_RobotApplication.Project.Save();
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Robot_Adapter/AdapterActions/Execute.cs:107]: yeah, fair enough.

So, if untitled project open, and save is true, then return an error and stop attempts of trying to exit?

[the remark this answers is not in the document]
Comment (peterjamesnugent) [reply Robot_Adapter/AdapterActions/Execute.cs:107]: Yeah I think return error and `false`.

The alternative would be no save and proceed to quit which could result in lost work.

**Review by peterjamesnugent — APPROVED** [review]
LGTM.

We will need to add a similar check for the `Close` command before it is saved - but that can be raised in a separate issue/PR.

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
noticed we get an error if `SaveBeforeClosing` is true but the file is not yet saved (due to being un-titled); in ETABS/SAP this is handled with a dialogue box from the software, so it's easy for a user to fix, but in Robot it's handled as a BHoM error. Not sure how this will affect testing (make sure bhombot sets the boolean to false, I suppose)

Otherwise, looks great.

**Comment by peterjamesnugent** [comment]
@JosefTaylor I handle this on Lusas by checking the database name, if it's empty then it returns an error message and `false`.

**Comment by IsakNaslundBh** [comment]
> Not sure how this will affect testing (make sure bhombot sets the boolean to false, I suppose)

The testing will call the command with false, see https://github.com/BHoM/Test_Toolkit/blob/5a9a870b19da08f202e95261c984a8b60eb53a0d/InteroperabilityTest_Engine/Compute/PushPullCompare.cs#L100

Agree there is some discrepancy between the two, which we may or may not want to remove, but the most important points to me for this command is.

1. If SaveBeforeClose is false, then the software should be exited, no matter what for the testing to work.
2. If SaveBeforeClose is true, the software should not close straight away without giving the user a chance to save the work if the command was not able to save it for you.

How 2. is achieved is to me at least of secondary importance. Etabs and SAP both had a boolean input to the exit method on the API that I made use of. Robot only had something half-similar, so opted for another solution.

If you disagree with this, I can try to find an aligning solution between the toolkits :)

**Review by alelom — DISMISSED** [review]
LGTM, only looked at the code.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Review by alelom — APPROVED** [review]
LGTM. Tested, works as expected.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check dataset-compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**